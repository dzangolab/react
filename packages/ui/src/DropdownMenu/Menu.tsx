import React, { memo } from "react";

export interface MenuItem {
  className?: string;
  disabled?: boolean;
  display?: boolean;
  key?: string;
  label?: string;
  onClick?: () => void;
}

export interface MenuProperties {
  className?: string;
  menu: MenuItem[];
  renderOption?: (value: MenuItem) => JSX.Element;
}

const Menu: React.FC<MenuProperties> = ({
  className,
  menu = [],
  renderOption,
}) => {
  return (
    <ul className={`dz-dropdown-menu ${className || ""}`.trimEnd()}>
      {menu.map(
        ({ className, disabled, onClick, display = true, ...item }, index) =>
          display ? (
            <li
              key={item.key || `menu-item-${index}`}
              onClick={disabled ? undefined : onClick}
              className={className}
              aria-disabled={disabled}
            >
              {renderOption ? (
                renderOption(item)
              ) : (
                <span className="dz-menu-item">{item.label}</span>
              )}
            </li>
          ) : null,
      )}
    </ul>
  );
};

export default memo(Menu);
