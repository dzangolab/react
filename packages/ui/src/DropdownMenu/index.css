.dropdown-container {
  cursor: pointer;
  align-items: center;
  display: grid;
}

.dropdown-container > ul.dropdown-menu {
  display: none;
  background-color: var(--dropdown-bg-color, #fff);
  border: var(--dropdown-border, 1px solid grey);
  border-radius: var(--dropdown-border-radius, 5px);
  list-style: none;
  padding-inline-start: 0;
  padding: 0.15rem 0;
  position: absolute;
  top: 70%;
  z-index: var(--dropdown-z-index, 999);
}

.dropdown-container[data-aria-expanded="true"]
  > ul.dropdown-menu[data-aria-hidden="false"] {
  display: block;
}

.dropdown-container > ul.dropdown-menu li {
  padding: 0.25rem 0.5rem;
}

.dropdown-container > ul.dropdown-menu li > a {
  color: inherit;
  text-decoration: none;
  height: 100%;
  display: flex;
  align-items: center;
}

span.icon {
  padding: 0 0.25rem;
}

.dropdown-container > ul.dropdown-menu li:hover {
  background-color: var(--dropdown-bg-color-hover, #dbe9fa);
}

.dropdown-container > .label-container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

@media screen and (max-width: 576px) {
  .dropdown-container {
    display: flex;
    flex-direction: column;
    flex: 1;
    padding: 0.8rem;
    align-items: unset;
  }

  .dropdown-container[data-aria-expanded="true"] {
    background-color: #f7f7f7;
    padding-bottom: 0;
  }

  .dropdown-container:active,
  .dropdown-container:focus {
    background-color: var(--dropdown-bg-color-hover, #dbe9fa);
  }

  .dropdown-container > ul.dropdown-menu {
    position: relative;
    border: none;
    background: inherit;
    padding: 0.8rem 0 0;
  }

  .dropdown-container > ul.dropdown-menu li {
    padding: 0 0.5rem;
    height: 40px;
    display: flex;
    align-items: center;
    background: inherit;
  }

  .dropdown-container > ul.dropdown-menu li:active {
    background-color: var(--dropdown-bg-color-hover, #dbe9fa);
  }

  .dropdown-container > ul.dropdown-menu li:hover {
    background-color: inherit;
  }
}
