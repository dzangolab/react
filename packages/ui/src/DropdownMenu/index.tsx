import React, { useEffect, useRef, useState } from "react";

import "./index.css";

export interface DropdownMenuProperties {
  className?: string;
  collapseIcon?: React.ReactNode;
  dropdownMenu: React.ReactNode;
  expandIcon?: React.ReactNode;
  label: React.ReactNode;
}

const DropdownMenu: React.FC<DropdownMenuProperties> = ({
  className = "",
  collapseIcon = <>&#9652;</>,
  dropdownMenu,
  expandIcon = <>&#9662;</>,
  label,
}) => {
  const [expanded, setExpanded] = useState(false);
  const navBarReference = useRef<HTMLDivElement | null>(null);

  let _className = "dropdown-container";

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (
        navBarReference.current &&
        !navBarReference.current.contains(event.target as Node)
      ) {
        setExpanded(false);
      }
    }

    document.addEventListener("click", handleClickOutside);

    return () => {
      document.removeEventListener("click", handleClickOutside);
    };
  }, []);

  if (className) {
    _className = className + " " + _className;
  }

  return (
    <div
      ref={navBarReference}
      className={_className}
      data-aria-expanded={expanded}
      onClick={() => setExpanded(!expanded)}
    >
      <div className="label-container">
        {typeof label === "string" ? (
          <span className="label">{label}</span>
        ) : (
          label
        )}
        <span className="icon">{expanded ? collapseIcon : expandIcon}</span>
      </div>
      {expanded && <ul className="dropdown-menu">{dropdownMenu}</ul>}
    </div>
  );
};

export default DropdownMenu;
