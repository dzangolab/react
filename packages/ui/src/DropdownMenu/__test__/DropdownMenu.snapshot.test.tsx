import { render } from "@testing-library/react";
import { expect, test } from "vitest";

import DropdownMenu from "..";

test("Component matches snapshot", () => {
  const menuItems = [
    {
      icon: "pi pi-lock",
      label: "user title",
    },
    {
      icon: "pi pi-user",
      label: "user profile",
    },
  ];

  const { container } = render(<DropdownMenu menu={menuItems} />);
  expect(container).toMatchSnapshot();
});
