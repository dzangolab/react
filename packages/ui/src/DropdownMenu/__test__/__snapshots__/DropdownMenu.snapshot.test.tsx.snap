// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<div>
  <div
    class="popup-container"
  >
    <div
      aria-controls="popup-content"
      aria-expanded="false"
      class="popup-trigger"
    >
      <span
        class="dz-dropdown-menu-trigger"
      >
        <i
          class="pi pi-ellipsis-h"
        />
        <i
          class="dz-dropdown-icon pi pi-angle-down"
        />
      </span>
    </div>
  </div>
</div>
`;
